src/invocation/InvocationService.ts

```diff
@@ -412,6 +456,36 @@ export class InvocationService {
         });
     }
 
+    fetchSchema(schemaId: Long): Promise<void> {
+        return this.schemaService.fetchSchema(schemaId);
+    }
+
+    // eslint-disable-next-line @typescript-eslint/ban-types
+    registerSchema(schema: Schema, clazz: Function | undefined): Promise<void> {
+        return this.schemaService.put(schema).then(() => {
+            this.serializationService.registerSchemaToClass(schema, clazz);
+        })
+    }
+
+    private fetchSchemaAndTryAgain(schemaId: Long, clientMessage: ClientMessage, invocation: Invocation): void {
+        // We need to fetch the schema to deserialize compact.
+        this.logger.trace(
+            'SchemaService', `Could not find schema id ${schemaId} locally, will search on the cluster`
+        );
+
+        this.fetchSchema(schemaId).then(() => {
+            // Reset nextFrame since we tried to parse the message once.
+            clientMessage.resetNextFrame();
+            invocation.eventHandler(clientMessage);
+        }, (err: Error) => {
+            if (err instanceof SchemaNotFoundError) {
+                this.fetchSchemaAndTryAgain(schemaId, clientMessage, invocation);
+            } else {
+                throw err;
```

Comment:
If you change `fetchSchemaAndTryAgain` to return a promise, then you can return this as an error.